include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+// alternatively, including winsock2.h before windows.h to prevent inclusion of the Winsock 1.1 winsock.h header from within
+// windows.h note: winsock2.h will include windows.h if not already included
+//
+// define _SECURE_SOCKET_TYPES_DEFINED_ at the project level to have access to SocketSecurity* functions in ws2tcpip.h
+//
+// define INCL_WINSOCK_API_TYPEDEFS at the project level to make function typedefs available across various networking headers
+// note, winsock2.h defaults is to not include function typedefs - but these can be necessary when supporting multiple OS versions
+//
+// Link libs for functions referenced in this file: ws2_32.lib, ntdll.lib, and Fwpuclnt.lib (for secure socket functions)
+
+#if !defined(_WINSOCK2API_) && defined(_WINSOCKAPI_)
+#error The Winsock 1.1 winsock.h header was included before the Winsock 2 winsock2.h header - this will cause compilation errors - define WIN32_LEAN_AND_MEAN to avoid winsock.h included by windows.h, or include winsock2.h before windows.h
+#endif
+
+// Including Winsock and networking headers in the below specific sequence
+// These headers have many intra-header dependencies, creating difficulties when needing access to various functions and types
+// This specific sequence should compile correctly to give access to all available functions and types
+#include <winsock2.h>
+#include <ws2def.h>
+#include <ws2ipdef.h>
+#include <mswsock.h>
+#include <mstcpip.h>
+#include <ws2tcpip.h>
+#include <windns.h>
+#include <iphlpapi.h>
+
+// wil headers
+#include "wistd_type_traits.h"
+#include "resource.h"
+
+// Callers who are using the STL by pulling in other STL headers
+// might not have included <string>, which is required in network.h when WIL_USE_STL is defined
+// ensuring it's included so callers don't have to include <string> just for network.h
+#if (WIL_USE_STL && defined(WIL_ENABLE_EXCEPTIONS)) || defined(WIL_DOXYGEN)
+#include <string>
+#endif
+
+namespace wil
+{
+// Functions and classes that support networking operations and structures
+namespace network
+{
+    // A type that calls WSACleanup on destruction (or reset()).
+    // WSAStartup must be called for the lifetime of all Winsock APIs (synchronous and asynchronous)
+    // WSACleanup will unload the full Winsock catalog - all the libraries - with the final reference
+    // which can lead to crashes if socket APIs are still being used after the final WSACleanup is called
+    using unique_wsacleanup_call = ::wil::unique_call<decltype(&::WSACleanup), ::WSACleanup>;
+
+    // Calls WSAStartup; returns an RAII object that reverts, the RAII object will resolve to bool 'false' if failed
+    WI_NODISCARD inline ::wil::network::unique_wsacleanup_call WSAStartup_nothrow() WI_NOEXCEPT
+    {
+        WSADATA unused_data{};
+        const auto error{::WSAStartup(WINSOCK_VERSION, &unused_data)};
+        LOG_IF_WIN32_ERROR(error);
+
+        ::wil::network::unique_wsacleanup_call return_cleanup{};

```

Comment:
You can also say `return unique_wsacleanup_call{ error != 0 };` here - unique_call(bool), when passed "true" arms the callback. Passing "false" returns an unarmed callback.